src/aggregator/aggregation/counter.go

```diff
@@ -72,6 +73,11 @@ func (c *Counter) Update(timestamp time.Time, value int64) {
 	if c.HasExpensiveAggregations {
 		c.sumSq += value * value
 	}
+
+	// Keep the last annotation which was set.
+	if len(annotation) > 0 {
+		c.annotation = annotation
```

Comment:
You'll actually want to copy this annotation since this comes from m3msg and the protobuf reuses the bytes when it decodes the next message in the m3msg server.

Also perhaps we should reuse memory here if we already have one.

Perhaps something like:

```go
if len(annotation) > 0 {
  if cap(c.annotation) < annotation) {
    c.annotation = make([]byte, 0, 2*annotation) // Twice as long in case another one comes in and we could avoid realloc as long as less than this first alloc.
  }
  c.annotation = append(c.annotation[:0], annotation...) // Reuse any previous allocation while taking a copy.
}
```

Reply:
You'll want to do this for gauge/timer too obviously.